### PR TITLE
[BEAM-2963] Preliminary refactors to PackageUtil

### PR DESCRIPTION
**Pull request by kennknowles**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

These are some basic refactors to give `PackageUtil` a little TLC, making it more usable and re-usable.

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @tgroh since I know you are looking at artifact-related things. This may or may not actually be interesting to you.

**Comment by kennknowles** [comment]
Failure to fetch hbase deps.

**Comment by kennknowles** [comment]
retest this please

**Comment by kennknowles** [comment]
run Dataflow ValidatesRunner

**Comment by kennknowles** [comment]
The tests all passed, then timeout occurred while building Python. I've bumped the timeout and will retest, but FYI this is g2g.

**Comment by kennknowles** [comment]
retest this please

**Comment by tgroh** [comment]
Run Build Pipeline

**Comment by tgroh** [comment]
Run Pipeline Job

**Comment by kennknowles** [comment]
run precommit pipeline java

**Comment by kennknowles** [comment]
retest this please

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java [hunk 1]
```diff
@@ -136,141 +130,150 @@ public int compare(PackageAttributes o1, PackageAttributes o2) {
...
+    } catch (Exception e) {
+      throw new RuntimeException(
+          String.format(
+              "Could not stage classpath element: %s to %s",
+              attributes.getSource(), attributes.getDestination().getLocation()),
+          e);
+    }
+  }
+
+  private StagingResult stagePackageBlithely(
```
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: While I really really like this name, I fear that it doesn't communicate what it does sufficiently precisely, given standard java jargon

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java [hunk 2]
```diff
@@ -136,141 +130,150 @@ public int compare(PackageAttributes o1, PackageAttributes o2) {
...
+          throw new IOException(errorMessage, e);
+        }
+        long sleep = backoff.nextBackOffMillis();
+        if (sleep == BackOff.STOP) {
+          // Rethrow last error, to be included as a cause in the catch below.
+          LOG.error("Upload failed, will NOT retry staging of classpath: {}", source, e);
+          throw e;
+        } else {
+          LOG.warn(
+              "Upload attempt failed, sleeping before retrying staging of classpath: {}",
```
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: 'staging of package'

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java [hunk 3]
```diff
@@ -288,45 +291,69 @@ private static void stageOnePackage(
...
-    List<ListenableFuture<?>> futures = new LinkedList<>();
-    for (final PackageAttributes attributes : packageAttributes) {
-      futures.add(executorService.submit(new Runnable() {
-        @Override
-        public void run() {
-          stageOnePackage(attributes, numUploaded, numCached, retrySleeper, createOptions);
-        }
-      }));
+    for (String classpathElement : classpathElements) {
+      DataflowPackage sourcePackage = new DataflowPackage();
```
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: Is the creation + value migration worth factoring into its own method?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java (hunk 1 above)
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: While I really like this name, I don't think it communicates why this is so blithe in standard java Jargon.

Comment (kennknowles) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: Suggestions? I settled on this suffix some while ago for lack of a better option.

Comment (kennknowles) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: "Crashily"?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java (hunk 2 above)
Comment (kennknowles) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: > **tgroh** wrote:
> 'staging of package'


Done.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java (hunk 3 above)
Comment (kennknowles) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: > **tgroh** wrote:
> Is the creation + value migration worth factoring into its own method?


TBH this should be moved out to the command line parsing. It has no place here. For now I'd rather leave it as-is (this bit is a code move)

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java (hunk 1 above)
Comment (kennknowles) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: > **kennknowles** wrote:
> "Crashily"?


Done. Split into tryStagePackage which is trivial straight-line code, tryStagePackageWithRetries that does just that loop, crashing if it hits something unrecoverable, and stagePackageSynchronously that just decorates the error message.

**Comment by kennknowles** [comment]
run dataflow validatesrunner

**Comment by kennknowles** [comment]
We've seen a couple jobs run, since the basic presubmit does that, but I'll wait to see a lot of jobs run, since if this breaks it is very bad.

**Comment by kennknowles** [comment]
All the jobs in the `ValidatesRunner` suite succeeded. It timed out in dependency analysis. Filed [BEAM-3024](https://issues.apache.org/jira/browse/BEAM-3024).

**Pull request merged**